ssm/variational.py

```diff
@@ -283,75 +334,138 @@ def sample_discrete_states(self):
+        negentropy = 0
+        for prms in self.params:
+            log_pi0 = np.log(prms["pi0"] + 1e-16) - logsumexp(prms["pi0"])
+            log_Ps = np.log(prms["Ps"] + 1e-16) - logsumexp(prms["Ps"],
+                                                            axis=1,
+                                                            keepdims=True)
+            (Ez, Ezzp1, normalizer) = hmm_expected_states(prms["pi0"],
+                                                          prms["Ps"],
+                                                          prms["log_likes"])
+
+            negentropy -= normalizer  # -log Z
+            negentropy += np.sum(Ez[0] * log_pi0)  # initial factor
+            negentropy += np.sum(Ez * prms["log_likes"])  # unitary factors
+            negentropy += np.sum(Ezzp1 * log_Ps)  # pairwise factors
+        return -negentropy
+
+    def _continuous_entropy(self, expectations):
+        negentropy = 0
+        Exs, ExxTs, ExxnTs, log_Zs = expectations
+        for prms, Ex, ExxT, ExxnT, log_Z in zip(self.params,
+                                                Exs,
+                                                ExxTs,
+                                                ExxnTs,
+                                                log_Zs):
+            # Pairwise terms
+            negentropy += np.sum(-0.5 * trace_product(prms["J_ini"], ExxT[0]))
+            negentropy += np.sum(-0.5 * trace_product(prms["J_dyn_11"], ExxT[:-1]))
+            negentropy += np.sum(-0.5 * trace_product(prms["J_dyn_22"], ExxT[1:]))
+            negentropy += np.sum(-0.5 * trace_product(prms["J_obs"], ExxT))
+            negentropy += np.sum(-1.0 * trace_product(prms["J_dyn_21"], ExxnT))
+
+            # Unary terms
+            negentropy += np.sum(prms["h_ini"] * Ex[0])
+            negentropy += np.sum(prms["h_dyn_1"] * Ex[:-1])
+            negentropy += np.sum(prms["h_dyn_2"] * Ex[1:])
+            negentropy += np.sum(prms["h_obs"] * Ex)
+
+            # Log normalizer
+            negentropy -= log_Z
+        return -negentropy
+
+    def _approx_continuous_entropy(self, sample):
+        """Return approximate entropy of the continuous posterior q(x) via sampling.
+
+        This can lead to a lower variance estimate of the gradient of ELBO. See
+        e.g "Sticking the Landing: Simple, Lower-Variance Gradient Estimators
+        for Variational Inference" https://arxiv.org/pdf/1703.09194.pdf
+        """
+        assert isinstance(sample, list) and len(sample) == len(self.datas)
+        negentropy = 0
+        for s, prms in zip(sample, self.params):
+            J_diag = prms["J_obs"]
+            J_lower_diag = prms["J_dyn_21"]
+            h = prms["h_obs"]
+            negentropy += block_tridiagonal_log_probability(
+                s,
+                J_diag,
+                J_lower_diag,
+                h
+            )
+        return -negentropy
+
+    def entropy(self, expectations=None, sample=None):
```

Comment:
I don't think the `entropy` function should take any arguments.  It can generate the expectations internally, and it won't cost anything if we cache the values.

Reply:
agreed. It won't take any arguments once the caching is set up.